extensions/amp-analytics/0.1/amp-analytics-3p-message-queue.js

```diff
@@ -0,0 +1,168 @@
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.3pMessageQueue';
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ */
+export class AmpAnalytics3pMessageQueue {
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!HTMLIFrameElement} frame The cross-domain iframe to send
+   * messages to
+   */
+  constructor(win, frame) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!HTMLIFrameElement} */
+    this.frame_ = frame;
+
+    /** @private {boolean} */
+    this.isReady_ = false;
+
+    /** @private {!../../../src/3p-analytics-common.AmpAnalytics3pEvent} */
+    this.creativeToPendingMessages_ = {};
+
+    /** @private
+     *  {!../../../src/3p-analytics-common.AmpAnalytics3pNewCreative} */
+    this.creativeToExtraData_ = {};
+
+    /** @private {string} */
+    this.messageType_ = this.frame_.getAttribute('data-amp-3p-sentinel') +
+        AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT;
+
+    /** @private {!../../../src/iframe-helper.SubscriptionApi} */
+    this.postMessageApi_ = new SubscriptionApi(this.frame_,
+        this.messageType_,
+        true,
+        () => {
+          this.setIsReady();
+        });
+  }
+
+  /**
+   * Returns whether the queue has been marked as ready yet
+   * @return {boolean}
+   * @VisibleForTesting
+   */
+  isReady() {
+    return this.isReady_;
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   * @VisibleForTesting
+   */
+  setIsReady() {
+    this.isReady_ = true;
+    this.flushQueue_();
+  }
+
+  /**
+   * Returns how many creativeId -> message(s) mappings there are
+   * @return {number}
+   * @VisibleForTesting
+   */
+  queueSize() {
+    return Object.keys(this.creativeToPendingMessages_).length;
+  }
+
+  /**
+   * Sets extra config data to be sent to a cross-domain iframe.
+   * @param {!string} creativeId Identifies which creative is sending the
+   * extra data
+   * @param {!string} extraData The extra config data
+   */
+  setExtraData(creativeId, extraData) {
+    dev().assert(!this.creativeToExtraData_[creativeId],
+        'Replacing existing extra data for ' + creativeId);
+    this.creativeToExtraData_[creativeId] = extraData;
+    this.flushQueue_();
+  }
+
+  /**
+   * Test method to get extra config data to be sent to a cross-domain iframe.
+   * @param {!string} creativeId Identifies which creative is sending the
+   * extra data
+   * @returns {string} The extra config data
+   */
+  getExtraData(creativeId) {
+    return this.creativeToExtraData_[creativeId];
+  }
+
+  /**
+   * Enqueues an AmpAnalytics3pEvent message to be sent to a cross-domain
+   * iframe.
+   * @param {!string} creativeId Identifies which creative is sending the message
+   * @param {!string} event The event to be enqueued and then sent to the iframe
+   */
+  enqueue(creativeId, event) {
```

Comment:
Generated in transport. Will rename as you suggested.
Note to self: TODO Saturday